### PR TITLE
feat: builder incentives

### PR DESCRIPTION
**Pull request by AlexD10S**
## Description 
Implement the mechanism for the Builder Incentives Protocol, designed to encourage smart contract developers by rewarding them based on the actual usage of their contracts over time.

The system operates in time periods called **eras**,  where the protocol tracks contract activity and distributes rewards.
This is being tracked using a signed extension `contract_fee_handler::ContractFeeHandler` which is an extension of the `pallet_transaction_payment::ChargeTransactionPayment<Runtime>` but tracking the `Contract::call` extrinsic.

Developers can register their contracts (dApps) with the `pallet-builder-incentives`, assigning a beneficiary account to receive rewards.

The runtime has been customized to include logic for handling fees, with a portion of the general fees directed to the `pallet-builder-incentives`. At the end of each era, during the `on_initialize` hook of the `pallet-builder-incentives`, era information is updated, and developers can claim their rewards for the completed era.

#### How the reward is calculated?
The pot receives a percentage of all fees generated in the system, along with any additional contributions made via the deposit_funds extrinsic. Contracts earn a share of the total reward pool for a given era, based on the proportion of fees they generated relative to the total fees from all registered contracts.

### Next Steps
- [ ] Ensure that only the contract owner (by querying the pallet revive) or the contract itself (via its H160 address) can call the register_contract extrinsic.
- [ ] Benchmarking and Weights. Special attention in the signed extension as it performs 2 read and 2 writes (in case is a contract call).
- [ ] Proper Fee Distribution: For the PoC 50% of fees are allocated to the protocol, and 50% are sent to the author. In the final version, fees should be distributed to collators, the Treasury, and maintainers and the builder incentives protocol based on predefined percentages.
- [ ] Registration Limits: Set a maximum limit for registered dApps, introduce a deposit requirement to prevent spam, and implement an unregistering mechanism.

### Additional Features to Explore
- Mechanism to update beneficiaries and maybe introduce contract managers. This could enable a market for trading reward rights (e.g., royalties), allowing individuals or groups to purchase the right to receive contract rewards.
- Introduce Claim Periods: Implement a claim period for developers to collect their rewards. If unclaimed after the period expires, the rewards could be returned to the pot for dApp incentives or transferred to the Treasury. This will prevent to have unaccesible DOTs and clean the state.

### CODE REVIEW

File: pallets/builder-incentives/src/types.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+use crate::Config;
+
+pub type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+/// AccountContractId for pallet_revive
+pub type AccountContractId = H160;
+/// Era number type
+pub type EraNumber = u32;
+/// An index to a block.
+pub type BlockNumber = u32;
```
Comment (evilrobot-01) [inline pallets/builder-incentives/src/types.rs:17]: Can get via existing BlockNumberFor<T> in frame

File: pallets/builder-incentives/src/types.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+/// Type aliases used for interaction with `OnChargeTransaction`.
+pub(crate) type OnChargeTransactionOf<T> =
+	<T as pallet_transaction_payment::Config>::OnChargeTransaction;
+
+/// OnChargeTransactionBalance type alias.
+pub(crate) type OnChargeTransactionBalanceOf<T> =
+	<OnChargeTransactionOf<T> as OnChargeTransaction<T>>::Balance;
+
+pub(crate) type ContractsOf<T> = <T as pallet_revive::Config>::Currency;
```
Comment (evilrobot-01) [inline pallets/builder-incentives/src/types.rs:27]: Weird type alias - contracts referring to currency associated type.

File: pallets/builder-incentives/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+		///
+		/// # Parameters
+		/// - `beneficiary`: The account that will be the beneficiary of the contract incentives.
+		/// - `contract`: The smart contract's account to be registered.
+		#[pallet::call_index(0)]
+		#[pallet::weight(Weight::zero())]
+		pub fn register_contract(
+			origin: OriginFor<T>,
+			beneficiary: T::AccountId,
+			contract: AccountContractId,
```
Comment (evilrobot-01) [inline pallets/builder-incentives/src/lib.rs:189]: This needs to be changed. The contract should be the origin, removing the need for this parameter.

File: runtime/devnet/src/lib.rs [hunk 4]
```diff
@@ -635,6 +693,10 @@ mod runtime {
 	#[runtime::pallet_index(150)]
 	pub type Fungibles = fungibles::Pallet<Runtime>;
 
+	// Builer Incentives
+	#[runtime::pallet_index(151)]
```
Comment (evilrobot-01) [inline runtime/devnet/src/lib.rs:697]: Needs updating as that index is now taken.

**Review by evilrobot-01 — CHANGES_REQUESTED** [review]
Two things need fixing imo - the pallet index in the runtime and the fact that anyone can register a contract, which should only be callable via the API.

The rest of the comments are minor and only included in case helpful. No changes are required for these.

File: pallets/builder-incentives/src/types.rs (hunk 2 above)
Comment (AlexD10S) [reply pallets/builder-incentives/src/types.rs]: You are right, no need for it.

**Review by evilrobot-01 — APPROVED** [review]

**Pull request merged**